### ISSUE TITLE
spack view fails saying share/info/dir already present

### ISSUE THREAD

**Issue by cpraveen** [spack/spack#6266]
I try to create symlink with spack view using this
```
SPACK_VIEW_DIR=$HOME/Applications/opt
ACTION=symlink

rm -rf $SPACK_VIEW_DIR
mkdir $SPACK_VIEW_DIR

spack view -v $ACTION $SPACK_VIEW_DIR wget
spack view -v $ACTION $SPACK_VIEW_DIR gcc
```
and this happens

```
==> [/Users/praveen/Applications/opt] Linked package: zlib@1.2.11%clang@9.0.0-apple+optimize+pic+shared arch=darwin-highsierra-x86_64 /t74siw3
==> [/Users/praveen/Applications/opt] Linked package: openssl@1.0.2k%clang@9.0.0-apple arch=darwin-highsierra-x86_64 /u7nfvs3
==> [/Users/praveen/Applications/opt] Linked package: wget@1.19.1%clang@9.0.0-apple~libpsl~pcre~python ssl=openssl +zlib arch=darwin-highsierra-x86_64 /aly5ysf
==> Warning: [/Users/praveen/Applications/opt] Skipping already linked package: zlib@1.2.11%clang@9.0.0-apple+optimize+pic+shared arch=darwin-highsierra-x86_64 /t74siw3
==> Error: [/Users/praveen/Applications/opt] Cannot link package gcc, file already exists: /Users/praveen/Applications/opt/share/info/dir
```

**Comment by davydden** [comment]
> file already exists

can you check if it does? Looks like a directory to me...

**Comment by cpraveen** [comment]
Yes it does. dir is just a file. Many packages seem to install it. So there is a conflict. So wget has linked it and contains

```
This is the file .../info/dir, which contains the
topmost node of the Info hierarchy, called (dir)Top.
The first time you invoke Info you start off looking at this node.
^_
File: dir,  Node: Top   This is the top of the INFO tree

  This (the Directory node) gives a menu of major topics.
  Typing "q" exits, "?" lists all Info commands, "d" returns here,
  "h" gives a primer for first-timers,
  "mEmacs<Return>" visits the Emacs manual, etc.

  In Emacs, you can click mouse button 2 on a menu item or cross reference
  to select it.

* Menu:

Network applications
* Wget: (wget).                 Non-interactive network downloader.
```

and then gcc also tries to link it.

**Comment by davydden** [comment]
Probably Spack could treat files in `share` prefix less strict. 

p.s. You certainly want to throw error if there is some conflict for libraries or headers, though.

**Comment by healther** [comment]
We found something similar, but with e.g.
```
etc/spack/linux-debian8-x86_64/gcc-4.9.2/llvm-4.0.0-zldwmrls63ma4ql2zg7fk7amjktl7mgd/lib/python2.7/site-packages/six.py
```

LLVM is particularly bad as it puts quite a number of files in non-llvm-only directories, but it isn't the only culprit (albeit I don't have a list of all conflicting packages that we found, at some point I just gave up)

As a workaround we currently use 
https://github.com/electronicvisions/spack/commit/ce2758a26d3e15b102eefa7818c1146ea4233055 
There are two important changes in that patch:

1. Activating a package in a view no longer tries to activate all dependencies (this wasn't necessary, as the list of specs to activate already contained all dependencies)
2. Adds an option ```ignore_conflicts``` to ```package.activate```, which downgrades the Error thrown to a Warning and then behaves like ```tree.merge(target, ignore_conflicts)```

So far we haven't found a conflict that wasn't (up to reformatting) identical. I briefly thought about doing a checksum comparison, but the above mentioned ```six.py``` was differently formatted and it's probably not a good idea to go down that rabbit hole to far...

I can polish this up and open a PR if that seems like a good solution, but the question of how spack SHOULD behave in case of conflicting files is one that should be discussed a bit more broadly

**Comment by scheibelp** [comment]
You can currently pass `--ignore-conflicts` to the `view` command like:

`spack view soft --ignore-conflicts`

> Activating a package in a view no longer tries to activate all dependencies (this wasn't necessary, as the list of specs to activate already contained all dependencies)

Although this is redundant, did view creation fail without this change?

**Comment by healther** [comment]
oh sorry, I should have prefaced my statement with I was on vacation and haven't followed #5415 too closely. Sorry for jumping the gun.

**Comment by cpraveen** [comment]
@scheibelp I dont see any option like ```--ignore-conflicts```.

I now dont create view for dependencies using ```-d no``` which reduces most conflicts. But still, I found that petsc and tbb where trying to link to same file ```include/index.html```. Since this is just some doc file, I manually delete this file to avoid the conflict.

Maybe an option like ```--ignore-share-conflicts``` would be useful. But this will not help to resolve conflicts inside ```include``` like the one above.

**Comment by obreitwi** [comment]
What is actually needed is generalized way to deal with specific files/directories when creating a view.

One example would be the merging if `easy_install.pth`-files for python extensions. Another is the dir file for `info`, the correct way to deal with this would be to call [install-info](https://linux.die.net/man/1/install-info) when adding and removing any spec to/from the view.

The two commands would be:
`install-info --info-dir=<path-to-view>/share/info <path-to-info-file-in-spec>.info`
`install-info --delete --info-dir=<path-to-view>/share/info <path-to-info-file-in-spec>.info`

**Comment by scheibelp** [comment]
> I dont see any option like `--ignore-conflicts`

@cpraveen

At the moment `--ignore-conflicts` will not be listed as an option if you do `spack view --help` but you can actually pass it. I should update the help for the command.

**Comment by JavierCVilla** [comment]
I have the same problem with files like `NOTICE` or `LICENSE`, even using the `--ignore-conflicts` option.

**Comment by scheibelp** [comment]
@JavierCVilla would you be able to provide the full command along with the output?

**Comment by cpraveen** [comment]
I now have the problem that more than one package is trying to install include/colamd.h which prevents creating views. Does this mean "spack views" are not a good way to work with spack ?

**Comment by healther** [comment]
> I now have the problem that more than one package is trying to install include/colamd.h which prevents creating views. Does this mean "spack views" are not a good way to work with spack ?

Yes and no. The problem is more fundamental than views themselves. What you see is the result of multiple packages having installed a `include/colamd.h` file. Which ever way you would install these packages you would end up with that conflict! Having system installs would mean that in the global include folder the second package would find the first package's `colamd.h` and would have to decide what to do about that.
So the conflict is inherent in the way these packages are installed and not "spack's fault" or "views" fault (i.e. you have the same situation if you use modules, then the first package in your environment will be used)

`spack view` by default errors out in case of such a conflict. In general this is too conservative, as especially for header files, multiple packages could ship identical files. However we ran into this when `llvm` was shipping its own `six.py` which was syntactically different from `py-six`'s `six.py` but semantically identical (i.e. they added line breaks). At that point we decided that we didn't want to go down that rabbit hole. 
In practice, just always ignoring all conflicts didn't blow up in our faces (so far).
